### ISSUE TITLE
GetServiceOrThrow can't get an implementation of an interface

### ISSUE THREAD

**Issue by DavidAmesPup** [agileobjects/AgileMapper#73]
When I call GetServiceOrThrow<> for an interface type, I get an "No service of type X is available"

However, calling GetServiceOrThrow for a concrete type that implements that interface succeeds.

` 
           
            public interface ISystemClock
            {
                DateTimeOffset Now { get; }
            }

            public class TestingClock : ISystemClock
            {
                public TestingClock(string now)
                {
                    Now = DateTimeOffset.Parse(now);
                }
                public DateTimeOffset Now { get; }
            }

            public class MyMappingConfiguration : MapperConfiguration
            {
                protected override void Configure()
                {
                    //This works
                    var xx = GetServiceOrThrow<TestingClock>();

                    //This blows up with "No service of type ISystemClock is available"
                    var systemClock = GetServiceOrThrow<ISystemClock>();
            }
        }

        var mapper = Mapper.CreateNew();
                    mapper.WhenMapping.UseConfigurations.From<MyMappingConfiguration>(new TestingClock("1-Jan-2017 2:00 am +11:00"));
                    _mapper.Map(myObject).ToANew<readModel>();
             
`

From looking at the documentation here: https://github.com/agileobjects/AgileMapper/wiki/Configuration-Classes#injecting-services it looks like this sort of thing is supported?

Cheers,
dave

**Comment by SteveWilkes** [comment]
Hi Dave!

At the moment services passed into `UseConfigurations` are registered against the concrete type of the object passed in - that's an oversight on my part - I'll update it to register interfaces too. Should have a new release by mid next week. Thanks for the feedback!

Steve

**Comment by SteveWilkes** [comment]
This issue has been made obsolete in the latest code. Services are now made available via first registering a service provider object - typically a DI container - and requests for services are then forwarded to the container and the result returned.

For example, using Microsoft's DI container:

```C#
// Add an ISystemClock implementation to the ServiceCollection:
services.AddSingleton<ISystemClock>(new TestingClock());

// Build an IServiceProvider instance from the ServiceCollection:
var provider = services.BuildServiceProvider();

// Register the provider as the source of services:
mapper.WhenMapping.UseServiceProvider(provider);

// In a MapperConfiguration, this now just calls GetService() 
// on the registered IServiceProvider:
var xx = GetService<ISystemClock>();

// ...and you can also retrieve the registered provider using:
var sp = GetServiceProvider<IServiceProvider>();
```

`GetService` and `GetServiceProvider` are also available to [custom data sources](/agileobjects/AgileMapper/wiki/Configuring-Member-Values), [callbacks](/agileobjects/AgileMapper/wiki/Configuring-Mapping-Callbacks) and [exception handlers](/agileobjects/AgileMapper/wiki/Configuring-Exception-Handling).

Apologies for the breaking API change - `GetServiceOrThrow` and `GetServiceOrNull` have been removed - but this does seem to me like the best way forward.

Release to follow, and I'll update the documentation once it's there. Thanks for the feedback! :)

**Comment by DavidAmesPup** [comment]
Cheers - it's actually a bit of a shame that we have to use service-containers to achieve this.

We have unit tests around our mapping functionality and having to configure a service container within our unit tests seems little heavy handed.

Having said that - I can appreciate that writing IoC container functionality is difficult and probably beyond the scope of AgileMapper so I see why you went down this path.

Dave

**Comment by SteveWilkes** [comment]
Just to add - the service provider support is kind of duck-typed - the provider object only needs to have a GetService(Type), GetInstance(Type) or Resolve(Type) method - so it would be pretty easy to throw together a stub provider for unit testing. Something like this:

```C#
public class StubServiceProvider
{
    private readonly Dictionary<Type, object> _services;

    public StubServiceProvider(params object[] services)
    {
        var servicesAndTypes = services
            .Select(s => new
            {
                Service = s,
                Types = new[] { s.GetType() }.Concat(s.GetType().GetInterfaces())
            });
        
        _services = new Dictionary<Type, object>();
        
        foreach (var serviceAndTypes in servicesAndTypes)
        {
            foreach (var type in serviceAndTypes.Types)
            {
                _services[type] = serviceAndTypes.Service;
            }
        }
    }
    
    public object GetService(Type serviceType)
    {
        return _services[serviceType];
    }
}
```

**Comment by SteveWilkes** [comment]
Released in [v0.26](https://www.nuget.org/packages/AgileObjects.AgileMapper/0.26.0) - including a `StubServiceProvider` class in the `AgileObjects.AgileMapper.Testing` namespace- documentation to follow. Thanks! :)

**Comment by SteveWilkes** [comment]
Service providers are now [documented](/agileobjects/AgileMapper/wiki/Dependency-Injection), including the [StubServiceProvider](/agileobjects/AgileMapper/wiki/Dependency-Injection#testing)
